File: Luogu/Tags/IDA/P2534.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
const int INF = 0x3f3f3f3f;
const int MOD = 1e9 + 7;
const int N = 100005;
#define endl '\n'

void solve() {
    int n; cin >> n;
    vector<int> A(n);
    for (auto &x : A) cin >> x;

    // 離散化
    auto B = A;
    sort(B.begin(), B.end());
    unordered_map<int, int> mp;
    for (int i = 0; i < n; i++) mp[B[i]] = i + 1;
    for (auto &x : A) x = mp[x];

    auto heuristic = [&](vector<int> &state) -> int {
        int cnt = 0;
        for (int i = 0; i < n - 1; i++)
            if (abs(state[i] - state[i + 1]) != 1) cnt++;
        if (state[n - 1] != n) cnt++;
        return cnt;
    };

    auto dfs = [&](auto &&dfs, int d, int max_d) -> bool {
        int h = heuristic(A);
        if (d + h > max_d) return false;
        if (h == 0) return true;
        for (int i = 1; i <= n; i++) {
            reverse(A.begin(), A.begin() + i);
            if (dfs(dfs, d + 1, max_d)) return true;
            reverse(A.begin(), A.begin() + i);
        }
        return false;
    };

    for (int max_d = heuristic(A); max_d <= (n - 1) * 2; max_d++) {
        if (dfs(dfs, 0, max_d)) {
            cout << max_d << endl;
            return;
        }
    }
    cout << -1 << endl;
    return;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    solve();
    return 0;
}

File: Luogu/Tags/IDA/P2534.py
"""
P2534 [AHOI2012] 铁盘整理
https://www.luogu.com.cn/problem/P2534

IDA* (Iterative Deepening A*)
"""
import sys
sys.setrecursionlimit(2000)

def solve():
    n = int(input())
    A = list(map(int, input().split()))

    # 離散化
    mp = {x: i + 1 for i, x in enumerate(sorted(A))}
    B = [mp[x] for x in A]

    def heuristic(state):
        cnt = 0
        for i in range(n - 1):
            if abs(state[i] - state[i + 1]) != 1:
                cnt += 1
        if state[-1] != n:
            cnt += 1
        return cnt

    def dfs(d, max_d, state) -> bool:
        h = heuristic(state)
        if d + h > max_d:  # Pruning
            return False
        if h == 0:
            return True
        for i in range(1, n):
            next_state = list(state[:i+1][::-1] + state[i+1:])
            if dfs(d + 1, max_d, next_state):
                return True
        return False

    max_d = heuristic(B)
    while max_d <= (n - 1) * 2:
        if dfs(0, max_d, B):
            print(max_d)
            break
        max_d += 1
    else:
        print(-1)

if __name__ == "__main__":
    solve()